[SourceKit/InterfaceGen] Also print the cross-import overlays of a module in its interface.

nathawes opened the pull request:
When printing the generated interface of a module, also print the decls from any underscored cross-import overlays it is the direct, or indirect underlying module of. Declarations are grouped by overlay, with a `MARK:` comment introducing each overlay, and a regular comment above each decl listing the required bystander modules that must be imported for the decl to be
available.

In addition, in each overlay:
- import declarations of the underlying modules are filtered out, since those modules are either other underscored cross-import overlays, or the target module the overlays are presented as being a part of.
- import declarations of modules that are also imported in the target module are filtered out, also because the overlay is presented as a conditional part of the target module.

@brentdax does that handling of import decls in overlays make sense?

Also update cursor info to report the owning module of the symbol under the cursor as the underlying/primary module when the symbol comes from a cross-import overlay.

Resolves rdar://problem/59445385

Commit:
[SourceKit/InterfaceGen] Also print the cross-import overlays of a module in its interface.

When printing the generated interface of a module, also print the decls from
any underscored cross-import overlays it is the direct, or indirect underlying
module of. Declarations are grouped by overlay, with a descriptive `MARK:`
comment introducing each overlay, and a regular comment above each decl listing
the required bystander modules that must be imported for the decl to be
available.

In addition in each overlay:
- import declarations of any underlying modules are filtered out, since they
  are either other underscored cross-import overlays, or the target module they
  are being presented as being part of.
- import declarations that are also in the target module are filtered out, since
  the overlay is being presented as a conditional part of the target module.

Resolves rdar://problem/59445385

nathawes commented:
@swift-ci please test

CodaFi commented:
@swift-ci test Linux

CodaFi commented:
@swift-ci test Windows

akyrtzi commented on test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.response:
```diff
@@ -0,0 +1,313 @@
+
+import B
+
+// Available when B is imported with A
```
How does this synthesized comment relate to a user documentation comment, could you add some doc-comments in the test cases?

nathawes replied:
The synthesized comment is printed before it at the moment. Added a test case showing the current behavior.

Commit:
[SourceKit] Update cursor info to report symbols from cross-import overlays as coming from the underylying module.

Also refactor some of the interface generation cross-import support code to be
shared.

Commit:
[SourceKit/InterfaceGen][test] Test that doc comments are printed after the required bystanders comment.

nathawes commented on lib/IDE/ModuleInterfacePrinting.cpp:
```diff
@@ -238,6 +238,295 @@ swift::ide::findGroupNameForUSR(ModuleDecl *M, StringRef USR) {
+/// Prints a single decl using the \p Printer and \p Options provided.
+///
+/// \returns Whether the given decl was printed.
+static bool printModuleInterfaceDecl(Decl *D,
```
This isn't new code - it's just pulled out of `printSubmoduleInterface()` below.

nathawes commented:
@swift-ci please test

akyrtzi commented on test/SourceKit/InterfaceGen/gen_swift_module_cross_import.swift.A.response:
```diff
@@ -0,0 +1,313 @@
+
+import B
+
+// Available when B is imported with A
```
Doesn't this seem a bit annoying to repeat it for every declaration? What do you think about only mention this once at the `MARK:` line?

nathawes replied:
@brentdax asked for it on every declaration in the radar so I'll redirect the question there. Brent, are you okay with Argyrios' suggestion?
It looks repetitive to me, but I'm also not sure how likely users are to scroll up in larger overlays if they get to one of these symbols by jump-to-definition or cmd+f.

akyrtzi replied:
> but I'm also not sure how likely users are to scroll up in larger overlays if they get to one of these symbols by jump-to-definition or cmd+f.
This is a good point, I retract my suggestion.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

nathawes commented:
@brentdax I'm going to go ahead and merge this so I can land a few subsequent PRs based on some the APIs added in this one, but it'd be great if you could take a look when you get the chance (I know you're probably busy build czar-ing). I'm not sure I've used the right terminology everywhere, and am probably missing some interesting test cases.
